File: ts-sdk/src/integrations/mcp-server.ts
/**
 * MCP (Model Context Protocol) server exposing agit operations as tools.
 *
 * Uses `@modelcontextprotocol/sdk` to register the following tools:
 *  - agit_commit
 *  - agit_log
 *  - agit_diff
 *  - agit_branch
 *  - agit_checkout
 *  - agit_merge
 *  - agit_revert
 *  - agit_status
 *
 * @example
 * ```ts
 * import { AgitClient } from "@agit/sdk";
 * import { createAgitMcpServer } from "@agit/sdk/integrations/mcp-server";
 *
 * const client = await AgitClient.open("/path/to/repo");
 * const server = createAgitMcpServer(client);
 * await server.listen(); // or transport of your choice
 * ```
 */

import { AgitClient } from "../client.js";
import { ActionType, MergeStrategy } from "../types.js";

// ---------------------------------------------------------------------------
// MCP SDK stubs – loaded at runtime so the import is optional.
// ---------------------------------------------------------------------------

interface McpServer {
  tool(
    name: string,
    description: string,
    schema: Record<string, unknown>,
    handler: (args: Record<string, unknown>) => Promise<McpToolResult>
  ): void;
  /** Start listening on the given transport. */
  connect(transport: unknown): Promise<void>;
}

interface McpToolResult {
  content: Array<{ type: "text"; text: string }>;
  isError?: boolean;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function ok(data: unknown): McpToolResult {
  return {
    content: [{ type: "text", text: JSON.stringify(data, null, 2) }],
  };
}

function fail(message: string): McpToolResult {
  return {
    content: [{ type: "text", text: `Error: ${message}` }],
    isError: true,
  };
}

function str(v: unknown, fallback = ""): string {
  return typeof v === "string" ? v : fallback;
}

function num(v: unknown, fallback: number): number {
  return typeof v === "number" ? v : fallback;
}

// ---------------------------------------------------------------------------
// Server factory
// ---------------------------------------------------------------------------

/**
 * Create and configure an MCP server exposing agit tools.
 *
 * The returned server object follows the `@modelcontextprotocol/sdk` `Server`
 * interface. Wire it to a transport (stdio, SSE, etc.) using `server.connect()`.
 */
export function createAgitMcpServer(client: AgitClient): McpServer {
  // Attempt to load the MCP SDK.
  let McpServerCtor: new (info: unknown, config: unknown) => McpServer;

  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const sdk = require("@modelcontextprotocol/sdk/server/index.js") as {
      Server: new (info: unknown, config: unknown) => McpServer;
    };
    McpServerCtor = sdk.Server;
  } catch {
    throw new Error(
      'MCP server requires "@modelcontextprotocol/sdk". ' +
        'Install it with: npm install @modelcontextprotocol/sdk'
    );
  }

  const server = new McpServerCtor(
    { name: "agit", version: "0.1.0" },
    { capabilities: { tools: {} } }
  );

  // ---- agit_commit --------------------------------------------------------
  server.tool(
    "agit_commit",
    "Commit the current agent state to the agit repository. Returns the commit hash.",
    {
      type: "object",
      properties: {
        memory: {
          description: "Agent working memory (arbitrary JSON).",
        },
        world_state: {
          description: "External world state snapshot (arbitrary JSON).",
        },
        message: {
          type: "string",
          description: "Commit message describing this checkpoint.",
        },
        action_type: {
          type: "string",
          description:
            "Action type: tool_call | llm_response | user_input | system_event | retry | rollback | merge | checkpoint",
          default: "checkpoint",
        },
        cost: {
          type: "number",
          description: "Cumulative cost (e.g. token spend) at this point.",
        },
      },
      required: ["memory", "message"],
    },
    async (args) => {
      try {
        const hash = await client.commit({
          memory: args["memory"],
          world_state: args["world_state"] ?? {},
          message: str(args["message"], "checkpoint"),
          action_type:
            (str(args["action_type"]) as ActionType) || ActionType.Checkpoint,
          cost: typeof args["cost"] === "number" ? args["cost"] : undefined,
        });
        return ok({ hash });
      } catch (e) {
        return fail(String(e));
      }
    }
  );

  // ---- agit_log -----------------------------------------------------------
  server.tool(
    "agit_log",
    "List commit history for a branch or HEAD.",
    {
      type: "object",
      properties: {
        branch: {
          type: "string",
          description: "Branch name to traverse. Defaults to HEAD.",
        },
        limit: {
          type: "number",
          description: "Maximum number of commits to return. Defaults to 20.",
          default: 20,
        },
      },
    },
    async (args) => {
      try {
        const commits = await client.log({
          branch: str(args["branch"]) || undefined,
          limit: num(args["limit"], 20),
        });
        return ok(commits);
      } catch (e) {
        return fail(String(e));
      }
    }
  );

  // ---- agit_diff ----------------------------------------------------------
  server.tool(
    "agit_diff",
    "Show the diff between two commits.",
    {
      type: "object",
      properties: {
        hash1: { type: "string", description: "Base commit hash." },
        hash2: { type: "string", description: "Target commit hash." },
      },
      required: ["hash1", "hash2"],
    },
    async (args) => {
      try {
        const diff = await client.diff(str(args["hash1"]), str(args["hash2"]));
        return ok(diff);
      } catch (e) {
        return fail(String(e));
      }
    }
  );

  // ---- agit_branch --------------------------------------------------------
  server.tool(
    "agit_branch",
    "Create a new branch at the given source commit or HEAD.",
    {
      type: "object",
      properties: {
        name: { type: "string", description: "New branch name." },
        from: {
          type: "string",
          description:
            "Source branch name or commit hash. Defaults to HEAD.",
        },
      },
      required: ["name"],
    },
    async (args) => {
      try {
        await client.branch({
          name: str(args["name"]),
          from: str(args["from"]) || undefined,
        });
        return ok({ created: str(args["name"]) });
      } catch (e) {
        return fail(String(e));
      }
    }
  );

  // ---- agit_checkout ------------------------------------------------------
  server.tool(
    "agit_checkout",
    "Checkout a branch or commit hash, restoring the stored agent state.",
    {
      type: "object",
      properties: {
        target: {
          type: "string",
          description: "Branch name or commit hash to check out.",
        },
      },
      required: ["target"],
    },
    async (args) => {
      try {
        const state = await client.checkout(str(args["target"]));
        return ok(state);
      } catch (e) {
        return fail(String(e));
      }
    }
  );

  // ---- agit_merge ---------------------------------------------------------
  server.tool(
    "agit_merge",
    "Merge a branch into the current branch.",
    {
      type: "object",
      properties: {
        branch: {
          type: "string",
          description: "Name of the branch to merge in.",
        },
        strategy: {
          type: "string",
          description:
            "Conflict resolution strategy: ours | theirs | three_way. Defaults to ours.",
          default: "ours",
        },
      },
      required: ["branch"],
    },
    async (args) => {
      try {
        const strategyStr = str(args["strategy"], "ours");
        const strategy =
          strategyStr === "theirs"
            ? MergeStrategy.Theirs
            : strategyStr === "three_way"
            ? MergeStrategy.ThreeWay
            : MergeStrategy.Ours;

        const hash = await client.merge({
          branch: str(args["branch"]),
          strategy,
        });
        return ok({ hash });
      } catch (e) {
        return fail(String(e));
      }
    }
  );

  // ---- agit_revert --------------------------------------------------------
  server.tool(
    "agit_revert",
    "Create a revert commit that restores the state from a previous commit.",
    {
      type: "object",
      properties: {
        hash: {
          type: "string",
          description: "Commit hash to revert to.",
        },
      },
      required: ["hash"],
    },
    async (args) => {
      try {
        const state = await client.revert(str(args["hash"]));
        return ok(state);
      } catch (e) {
        return fail(String(e));
      }
    }
  );

  // ---- agit_status --------------------------------------------------------
  server.tool(
    "agit_status",
    "Return repository status: HEAD, current branch, and all branches.",
    { type: "object", properties: {} },
    async () => {
      try {
        return ok(client.status());
      } catch (e) {
        return fail(String(e));
      }
    }
  );

  return server;
}

File: python/agit/integrations/mcp_server.py
"""MCP Server – exposes agit tools via FastMCP."""
from __future__ import annotations

from typing import Any, Optional

from agit.engine.executor import ExecutionEngine

try:
    from fastmcp import FastMCP  # type: ignore[import]

    _FASTMCP_AVAILABLE = True
except ImportError:
    _FASTMCP_AVAILABLE = False

    class FastMCP:  # type: ignore[no-redef]
        """Minimal FastMCP stub."""

        def __init__(self, name: str) -> None:
            self.name = name
            self._tools: dict[str, Any] = {}

        def tool(self, fn: Any = None, **kwargs: Any) -> Any:
            if fn is not None:
                self._tools[fn.__name__] = fn
                return fn

            def decorator(f: Any) -> Any:
                self._tools[f.__name__] = f
                return f

            return decorator

        def run(self, transport: str = "stdio") -> None:
            pass


def create_mcp_server(
    engine: ExecutionEngine,
    server_name: str = "agit",
) -> FastMCP:
    """Build and return a FastMCP server with all agit tools registered.

    Parameters
    ----------
    engine:
        An initialised :class:`ExecutionEngine`.
    server_name:
        Name reported to MCP clients.

    Returns
    -------
    FastMCP:
        Ready-to-run MCP server. Call ``.run()`` to start serving.

    Example::

        engine = ExecutionEngine("./repo", agent_id="mcp-agent")
        server = create_mcp_server(engine)
        server.run()   # starts stdio MCP server
    """
    mcp = FastMCP(server_name)

    # ------------------------------------------------------------------
    # Tool: agit_init
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_init(repo_path: str = ".", agent_id: str = "mcp") -> dict[str, Any]:
        """Initialize an agit repository at the given path."""
        try:
            ExecutionEngine(repo_path=repo_path, agent_id=agent_id)
            return {"ok": True, "repo_path": repo_path}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_commit
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_commit(
        message: str,
        state: Optional[dict[str, Any]] = None,
        action_type: str = "checkpoint",
    ) -> dict[str, Any]:
        """Commit the given state (or current state) with a message."""
        try:
            s = state or engine.get_current_state() or {}
            h = engine.commit_state(s, message, action_type)
            return {"ok": True, "hash": h, "message": message}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_log
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_log(limit: int = 10) -> dict[str, Any]:
        """Return the commit history."""
        try:
            commits = engine.get_history(limit)
            return {"ok": True, "commits": commits}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_diff
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_diff(hash1: str, hash2: str) -> dict[str, Any]:
        """Compute the diff between two commit hashes."""
        try:
            d = engine.diff(hash1, hash2)
            return {"ok": True, "diff": d}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_branch
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_branch(
        name: Optional[str] = None,
        from_ref: Optional[str] = None,
    ) -> dict[str, Any]:
        """Create a branch (if name given) or list all branches."""
        try:
            if name:
                engine.branch(name, from_ref=from_ref)
                return {"ok": True, "created": name}
            return {"ok": True, "branches": engine.list_branches()}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_checkout
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_checkout(target: str) -> dict[str, Any]:
        """Checkout a branch or commit hash."""
        try:
            state = engine.checkout(target)
            return {"ok": True, "target": target, "state": state}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_merge
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_merge(branch: str, strategy: str = "three_way") -> dict[str, Any]:
        """Merge a branch into HEAD."""
        try:
            h = engine.merge(branch, strategy=strategy)
            return {"ok": True, "merge_commit": h}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_revert
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_revert(commit_hash: str) -> dict[str, Any]:
        """Revert to the state at a previous commit hash."""
        try:
            state = engine.revert(commit_hash)
            return {"ok": True, "reverted_to": commit_hash, "state": state}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_status
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_status() -> dict[str, Any]:
        """Return current repository status."""
        try:
            history = engine.get_history(1)
            return {
                "ok": True,
                "branch": engine.current_branch(),
                "branches": engine.list_branches(),
                "last_commit": history[0] if history else None,
            }
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_audit
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_audit(limit: int = 20) -> dict[str, Any]:
        """Return the audit log."""
        try:
            logs = engine.audit_log(limit)
            return {"ok": True, "entries": logs}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_state_replay
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_state_replay(commit_hash: str) -> dict[str, Any]:
        """Get the agent state at a specific commit without changing HEAD."""
        try:
            # Get state at commit via checkout and restore
            current = engine.current_branch() or "main"
            state = engine.checkout(commit_hash)
            engine.checkout(current)  # restore position
            return {"ok": True, "hash": commit_hash, "state": state}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    # ------------------------------------------------------------------
    # Tool: agit_search
    # ------------------------------------------------------------------
    @mcp.tool()
    def agit_search(
        query: str,
        action_type: Optional[str] = None,
        limit: int = 20,
    ) -> dict[str, Any]:
        """Search commits by message content or action type."""
        try:
            commits = engine.get_history(limit * 5)  # over-fetch for filtering
            results = []
            query_lower = query.lower()
            for c in commits:
                msg = c.get("message", "").lower()
                at = c.get("action_type", "")
                if query_lower in msg or query_lower in at:
                    if action_type and at != action_type:
                        continue
                    results.append(c)
                    if len(results) >= limit:
                        break
            return {"ok": True, "results": results, "count": len(results)}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    return mcp


def main() -> None:
    """CLI entry-point: start the MCP server."""
    import argparse

    parser = argparse.ArgumentParser(description="agit MCP server")
    parser.add_argument("--repo", default=".", help="Repository path")
    parser.add_argument("--agent", default="mcp", help="Agent ID")
    parser.add_argument("--transport", default="stdio", choices=["stdio", "sse"])
    parser.add_argument("--host", default="0.0.0.0", help="SSE host (default: 0.0.0.0)")
    parser.add_argument("--port", type=int, default=8000, help="SSE port (default: 8000)")
    args = parser.parse_args()

    engine = ExecutionEngine(repo_path=args.repo, agent_id=args.agent)
    server = create_mcp_server(engine)

    if args.transport == "sse":
        server.run(transport="sse", host=args.host, port=args.port)
    else:
        server.run(transport="stdio")


if __name__ == "__main__":
    main()
